### PR TITLE
Filter resource warnings in integration tests (#1825)

### PR DESCRIPTION
**Pull request by amarjandu**
see #1825 for discussion

### CODE REVIEW

File: test/azul_test_case.py [hunk 1]
```diff
@@ -3,12 +3,21 @@
...
 from botocore.credentials import Credentials
 import botocore.session
 
 
+class AzulTestCase(TestCase):
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        warnings.filterwarnings('ignore', 'unclosed.*<ssl.SSLSocket.*>', ResourceWarning)
```
Comment (jessebrennan) [inline test/azul_test_case.py:18]: I think this line makes more sense in the second commit. First commit refactors, second commit adds changes.

Comment (amarjandu) [reply test/azul_test_case.py:18]: Done

**Comment by jessebrennan** [comment]
Could you change the second commit title to "Filter resource warnings in integration tests (#1825)"? I forgot to comment that in my review, but once it's done, I approve.

File: test/azul_test_case.py [hunk 2]
```diff
@@ -3,12 +3,20 @@
 from unittest.mock import (
     patch,
 )
+import warnings
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:6]: Not in 1st commit

File: test/azul_test_case.py [hunk 3]
```diff
@@ -3,12 +3,20 @@
...
 
 import boto3.session
 from botocore.credentials import Credentials
 import botocore.session
 
 
+class AzulTestCase(TestCase):
+
+    @classmethod
+    def setUpClass(cls) -> None:
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:16]: Not in 1st commit.

**Comment by amarjandu** [comment]
> Changes look good. Their just not split correctly over the two commits. Either squash the commits or split them correctly.

@hannes-ucsc i opted to squash the commits down.

File: test/azul_test_case.py (hunk 1 above)
Comment (hannes-ucsc) [inline test/azul_test_case.py:18]: Every action in `setUp` should be matched by an inverse action in `tearDown`. IOW, `tearDown` should leave the world in the state that `setUp` was called in.

You could use `warning.resetwarnings()` but then you would have to assert in `setUp` that the warnings are indeed at the reset state. Luckily, there is `warnings.catch_warnings` which was made exactly for this purpose. You can even assert in `tearDown` that only expected warnings occurred. You can't use `warnings.catch_warnings` in conjunction with a `with` statement because a statement can't be split over two methods, but you can call `__enter__` and `__exit__` explicitly. Be sure to read the [PEP on that](https://www.python.org/dev/peps/pep-0343/#specification-the-with-statement). Plus there are some fixtures in our tests that use that technique already.

Comment (hannes-ucsc) [reply test/azul_test_case.py:21]: Now that we use the `catch_warnings` context manager we can actually assert that we only get a specific set of warnings. It's possible that we get no warnings or a subset of the set of expected warnings but the test should fail if we get an unexpected warning.

https://docs.python.org/3/library/warnings.html#testing-warnings

That's what I meant by 

> You can even assert in tearDown that only expected warnings occurred.

in my previous comment on this.

Comment (amarjandu) [reply test/azul_test_case.py:21]: @hannes-ucsc is there a list of Exceptions that we would permit to pass through?

Comment (amarjandu) [reply test/azul_test_case.py:21]: During an IT with a filter of: 

`warnings.filterwarnings('ignore', 'unclosed.*<ssl.SSLSocket.*>', ResourceWarning)`

the catch_warning record shows multiple instances of :

`{message : DeprecationWarning('Call to deprecated method fetch_bundle_manifest.'), category : 'DeprecationWarning', filename : '/Users/amar/dev/hca/azul/src/azul/azulclient.py', lineno : 92, line : None}`

File: test/azul_test_case.py [hunk 4]
```diff
@@ -1,30 +1,57 @@
...
-        cls._warning_state.__enter__()
-        warnings.filterwarnings('ignore', 'unclosed.*<ssl.SSLSocket.*>', ResourceWarning)
+        cls._warning_state = warnings.catch_warnings(record=True)
+        cls._warning_log = cls._warning_state.__enter__()
 
     @classmethod
     def tearDownClass(cls) -> None:
+        if cls._inspect_for_bad_warnings():
+            pass
+            # at this point the test should already have a status failed or pass, how  to modify the test that occured?
```
Comment (amarjandu) [inline test/azul_test_case.py:34]: @hannes-ucsc attempting to fail at this point does not seem to work as expected due to the result of the test already being determined.
Is there a way to modify the result of the testcase that has already finished?

File: test/azul_test_case.py (hunk 1 above)
Comment (amarjandu) [reply test/azul_test_case.py]: Decided to remove filterwarnings, and inspect all warnings caught by the `catch_warnings` these WarningMessages are to be compared against the permitted warnings, and regular expressions that are defined to finely filter messages we expect to see. 
We have to inspect ever single warning, and need to ensure that ALL warnings that fail to meet the filter requirements are displayed.

File: test/azul_test_case.py [hunk 5]
```diff
@@ -1,13 +1,54 @@
...
 from botocore.credentials import Credentials
 import botocore.session
 
+log = logging.getLogger(__name__)
+
+
+class AzulTestCase(TestCase):
+    _warning_state = None
+    _warning_log = List[warnings.WarningMessage]
+    metadata_api_warning_messages = [
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:20]: Inline. No advantage in extracting an attribute.

File: test/azul_test_case.py [hunk 6]
```diff
@@ -1,13 +1,54 @@
...
+        cls._warning_log = cls._warning_state.__enter__()
+        warnings.filterwarnings('ignore', '.*<ssl.SSLSocket.*>', ResourceWarning)
+        warnings.filterwarnings('ignore', '.*<socket.socket.*>', ResourceWarning)
+        warnings.filterwarnings('ignore', '.*Call to deprecated method fetch_bundle_manifest.*', DeprecationWarning)
+        for message_ in cls.metadata_api_warning_messages:
+            warnings.filterwarnings('ignore', message_, DeprecationWarning)
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        if len(cls._warning_log) > 0:
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:45]:
```suggestion
        if cls._warning_log:
```

But really just `self.assertFalse(cls._warning_log)`. The backtrace will contain all the unexpected warnings.

File: test/azul_test_case.py [hunk 7]
```diff
@@ -1,13 +1,54 @@
...
+            for warning in cls._warning_log:
+                log.error(log.error('Unexpected warning found: %s', warning))
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:47]:
```suggestion
                log.error('Unexpected warning found: %s', warning)
```

Was this even tested?

File: test/azul_test_case.py [hunk 8]
```diff
@@ -1,13 +1,54 @@
...
 import boto3.session
 from botocore.credentials import Credentials
 import botocore.session
 
+log = logging.getLogger(__name__)
+
+
+class AzulTestCase(TestCase):
+    _warning_state = None
+    _warning_log = List[warnings.WarningMessage]
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:19]:
```suggestion
    _warning_log: List[warnings.WarningMessage]
```

File: test/service/test_cart_export_endpoints.py [hunk 9]
```diff
@@ -46,9 +46,9 @@ def test_unified_endpoint_initiate_ok(self):
                     response = requests.post(export_url,
                                              headers=dict(Authorization=f'Bearer {test_jwt}'),
                                              allow_redirects=False)
-        self.assertEquals(301, response.status_code)
-        self.assertEquals(f'{export_url}?token={expected_export_token}',
-                          response.headers['Location'])
+        self.assertEqual(301, response.status_code)
```
Comment (hannes-ucsc) [inline test/service/test_cart_export_endpoints.py:49]: Now that you went there, these cosmetic changes should be in a separate commit.

File: test/test_app_logging.py [hunk 10]
```diff
@@ -85,3 +90,23 @@ def fail():
...
+        warnings.warn(" unclosed <ssl.SSLSocket fd=30, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM,"
+                      " proto=0, laddr=('192.168.1.11', 63179), raddr=('172.217.5.112', 443)>",
+                      category=ResourceWarning)
+        warnings.warn('Call to deprecated method fetch_bundle_manifest.)',
+                      category=DeprecationWarning)
+        warnings.warn("<class 'humancellatlas.data.metadata.api.LibraryPreparationProcess'> is deprecated",
+                      category=DeprecationWarning)
+
+    def test_unexpected_warning(self):
+        class TestUnexpectedWarnings(AzulUnitTestCase):
```
Comment (hannes-ucsc) [inline test/test_app_logging.py:107]: This isn't doing anything.

File: test/azul_test_case.py (hunk 7 above)
Comment (amarjandu) [reply test/azul_test_case.py:47]: Yeah the output looks like:
```
2020-06-22 18:41:40,158 ERROR   MainThread: Unexpected warning found: {message : DeprecationWarning("<class 'humancellatlas.data.metadata.api.LibraryPreparationProcess'> is deprecated"), category : 'DeprecationWarning', filename : '/Users/amar/dev/hca/azul/test/test_app_logging.py', lineno : 105, line : None}
```

Used it to find all the required filter patterns, fixed the extra log.error().

File: test/service/test_cart_export_endpoints.py (hunk 9 above)
Comment (amarjandu) [reply test/service/test_cart_export_endpoints.py:49]: @hannes-ucsc these changes were done  in bab2df07001bb31eb4dc7e54ac5d198524d332e9 do i need to change something?

File: test/test_app_logging.py (hunk 10 above)
Comment (amarjandu) [reply test/test_app_logging.py:107]: fixed

File: test/azul_test_case.py (hunk 6 above)
Comment (amarjandu) [reply test/azul_test_case.py:45]: @hannes-ucsc
I removed the length inspection.

 Would there be other modifications required to use the `self` reference in a `classmethod`?

I like the idea of relying on the assertFalse perform the inspection and the output of the warnings. I'm a little bit confused however; wouldn't assertFalse require the context of a test to be running?

File: test/test_app_logging.py [hunk 11]
```diff
@@ -85,3 +93,40 @@ def fail():
...
+                              category=ResourceWarning)
+
+        test_case = TU('test')
+        test_suite = TestSuite()
+        test_result = TestResult()
+        test_suite.addTest(test_case)
+        test_suite.run(test_result)
+
+        self.assertEqual(test_result.testsRun, 1)
+        self.assertEqual(one(test_result.errors)[0].description,
```
Comment (amarjandu) [inline test/test_app_logging.py:128]: @hannes-ucsc during our discussion today we talked about usage of one() being applicable to the test_result (we only wanted 1 test to be run here), however we did not note that one() requires an `iterate` to be passed as an argument.

File: test/azul_test_case.py [hunk 12]
```diff
@@ -1,14 +1,47 @@
...
+class AzulTestCase(TestCase):
+    _warning_state = None
+    _warning_log: List[warnings.WarningMessage]
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        cls._warning_state = warnings.catch_warnings(record=True)
+        cls._warning_log = cls._warning_state.__enter__()
+        permitted_warnings_ = {ResourceWarning: ['.*<ssl.SSLSocket.*>', '.*<socket.socket.*>'],
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:23]: See contribution guide on formatting literals.

File: test/service/test_cart_export_endpoints.py (hunk 9 above)
Comment (hannes-ucsc) [reply test/service/test_cart_export_endpoints.py:49]: No, I missed you already had a separate commit for this.

File: test/test_app_logging.py [hunk 13]
```diff
@@ -85,3 +93,40 @@ def fail():
                             azul_log.output[1],
                             'INFO:azul.chalice:Returning 500 response. To log headers and body, set AZUL_DEBUG to 1.'
                         )
+
+
+class TestPermittedWarnings(AzulUnitTestCase):
+
+    def test_permitted_warnings(self):
+        warnings.warn(" unclosed <ssl.SSLSocket fd=30, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM,"
```
Comment (hannes-ucsc) [inline test/test_app_logging.py:101]: I think one of these would be sufficient.

Comment (hannes-ucsc) [reply test/test_app_logging.py:101]: And a comment should explain why there are no assertions. That tearDownClass would raise if these one of these slipped through.

File: test/test_app_logging.py (hunk 11 above)
Comment (hannes-ucsc) [reply test/test_app_logging.py:128]: What's an `iterate`? This looks ok to me.

File: test/test_app_logging.py [hunk 14]
```diff
@@ -85,3 +93,40 @@ def fail():
...
+        test_case = TU('test')
+        test_suite = TestSuite()
+        test_result = TestResult()
+        test_suite.addTest(test_case)
+        test_suite.run(test_result)
+
+        self.assertEqual(test_result.testsRun, 1)
+        self.assertEqual(one(test_result.errors)[0].description,
+                         'tearDownClass (test_app_logging.TestUnexpectedWarnings.test_unexpected_warning.<locals>.TU)')
+        error_line = one(test_result.errors)[1].split('\n')[-2]
```
Comment (hannes-ucsc) [inline test/test_app_logging.py:130]: `one(test_result.errors)` should be extracted into separate variable. Or better, use tuple unpacking assignment to get at both tuple elements at the same time.

File: test/test_app_logging.py [hunk 15]
```diff
@@ -85,3 +93,40 @@ def fail():
...
+        self.assertRegex(error_line, "^AssertionError")
+        self.assertRegex(error_line, msg)
```
Comment (hannes-ucsc) [inline test/test_app_logging.py:132]:
```suggestion
        self.assertIn(msg, error_line)
```

File: test/test_app_logging.py [hunk 16]
```diff
@@ -85,3 +93,40 @@ def fail():
...
+        warnings.warn("<class 'humancellatlas.data.metadata.api.LibraryPreparationProcess'> is deprecated",
+                      category=DeprecationWarning)
+
+
+class TestUnexpectedWarnings(TestCase):
+
+    def test_unexpected_warning(self):
+        msg = "Testing unexpected warnings, nothing to see here."
+
+        class TU(AzulUnitTestCase):
```
Comment (hannes-ucsc) [inline test/test_app_logging.py:115]:
```suggestion
        class Test(AzulUnitTestCase):
```

File: test/test_app_logging.py [hunk 17]
```diff
@@ -85,3 +93,40 @@ def fail():
...
+
+            def test(self):
+                warnings.warn(message=msg,
+                              category=ResourceWarning)
```
Comment (hannes-ucsc) [inline test/test_app_logging.py:119]: one line

File: test/azul_test_case.py [hunk 18]
```diff
@@ -1,14 +1,47 @@
...
 )
+import warnings
 
 import boto3.session
 from botocore.credentials import Credentials
 import botocore.session
 
 
+class AzulTestCase(TestCase):
+    _warning_state = None
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:15]:
```suggestion
    _catch_warnings = None
```

File: test/azul_test_case.py [hunk 19]
```diff
@@ -1,14 +1,47 @@
...
+    _warning_log: List[warnings.WarningMessage]
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:16]:
```suggestion
    _caught_warnings: List[warnings.WarningMessage]
```

File: test/azul_test_case.py [hunk 20]
```diff
@@ -1,14 +1,47 @@
...
+                                                    "ProjectContact.contact_name is deprecated",
+                                                    "File.file_format is deprecated",
+                                                    "ProjectPublication.publication_title is deprecated",
+                                                    "ProjectPublication.publication_url is deprecated",
+                                                    "CellLine.cell_line_type is deprecated",
+                                                    ".*humancellatlas.data.metadata.api.DissociationProcess",
+                                                    ".*humancellatlas.data.metadata.api.EnrichmentProcess",
+                                                    ".+humancellatlas.data.metadata.api.LibraryPreparationProcess",
+                                                    ".*humancellatlas.data.metadata.api.SequencingProcess"]}
+        for warning_class, filter_messages in permitted_warnings_.items():
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:34]:
```suggestion
        for warning_class, message_patterns in permitted_warnings_.items():
```

File: test/azul_test_case.py [hunk 21]
```diff
@@ -1,14 +1,47 @@
...
+            for message in filter_messages:
```
Comment (hannes-ucsc) [inline test/azul_test_case.py:35]:
```suggestion
            for message_pattern in filter_messages:
```

File: test/test_app_logging.py (hunk 11 above)
Comment (amarjandu) [reply test/test_app_logging.py:128]: typo: `Iterable`
